File: src/Q48.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
  int lengthOfLongestSubstring(string s) {
    vector<int> pos = vector(128, -1);
    int l = 0, r = 0, ret = 0;
    int len = s.length();
    for (int i = 0; i < len; ++i) {
      int c = s[i];
      r += 1;
      if (pos[c] == -1) {
        ret = max(ret, r - l);
      } else {
        for (; l <= pos[c]; ++l) {
          // cout << s[l] << endl;
          pos[s[l]] = -1;
        }
      }
      pos[c] = i;
    }
    return ret;
  }
};

int main() {
  string s = "tmmzuxt";
  // string s = "bbtablud";
  Solution a;
  cout << a.lengthOfLongestSubstring(s) << endl;
  return 0;
}

File: src/Q48.rs
struct Solution;

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        use std::cmp::max;
        let mut exist = vec![false; 128];
        let mut l = 0;
        let mut r = 0;
        let bytes = s.as_bytes();
        let mut ret = 0;
        for c in bytes {
            while exist[*c as usize] {
                exist[bytes[l as usize] as usize] = false;
                l += 1;
            }
            exist[*c as usize] = true;
            r += 1;
            ret = max(ret, r - l);
        }
        ret
    }
}
